builtin/logical/transit/path_sign_verify.go

```diff
@@ -222,63 +261,175 @@ func (b *backend) pathSignWrite(ctx context.Context, req *logical.Request, d *fr
+				return logical.ErrorResponse(response[0].Error), response[0].err
+			}
+			return nil, response[0].err
+		}
+		resp.Data = map[string]interface{}{
+			"signature": response[0].Signature,
+		}
+		if len(response[0].PublicKey) > 0 {
+			resp.Data["public_key"] = response[0].PublicKey
+		}
 
-	if len(sig.PublicKey) > 0 {
-		resp.Data["public_key"] = sig.PublicKey
 	}
 
 	p.Unlock()
 	return resp, nil
 }
 
 func (b *backend) pathVerifyWrite(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	sig := d.Get("signature").(string)
-	hmac := d.Get("hmac").(string)
+	batchInputRaw := d.Raw["batch_input"]
+	var batchInputItems []batchRequestSignItem
+	if batchInputRaw != nil {
+		err := mapstructure.Decode(batchInputRaw, &batchInputItems)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to parse batch input: {{err}}", err)
+		}
+
+		if len(batchInputItems) == 0 {
+			return logical.ErrorResponse("missing batch input to process"), logical.ErrInvalidRequest
+		}
+	} else {
+		// use empty string if input is missing - not an error
+		inputB64 := d.Get("input").(string)
+
+		batchInputItems = make([]batchRequestSignItem, 1)
+		batchInputItems[0] = batchRequestSignItem{
+			"input": inputB64,
+		}
+		if sig, ok := d.GetOk("signature"); ok {
+			batchInputItems[0]["signature"] = sig.(string)
+		}
+		if hmac, ok := d.GetOk("hmac"); ok {
+			batchInputItems[0]["hmac"] = hmac.(string)
+		}
+		batchInputItems[0]["context"] = d.Get("context").(string)
+	}
+
+	// For simplicity, 'signature' and 'hmac' cannot be mixed across batch_input elements.
+	// If one batch_input item is 'signature', they all must be 'signature'.
+	// If one batch_input item is 'hmac', they all must be 'hmac'.
+	sigFound := false
+	hmacFound := false
+	missing := false
+	for _, v := range batchInputItems {
+		either := false
+		if _, ok := v["signature"]; ok {
+			sigFound = true
+			either = true
+		}
+		if _, ok := v["hmac"]; ok {
+			hmacFound = true
+			either = true
+		}
+		if !either {
+			missing = true
+		}
+	}
+
 	switch {
-	case sig != "" && hmac != "":
+	case batchInputRaw == nil && sigFound && hmacFound:
 		return logical.ErrorResponse("provide one of 'signature' or 'hmac'"), logical.ErrInvalidRequest
 
-	case sig == "" && hmac == "":
+	case batchInputRaw == nil && !sigFound && !hmacFound:
 		return logical.ErrorResponse("neither a 'signature' nor an 'hmac' were given to verify"), logical.ErrInvalidRequest
 
-	case hmac != "":
-		return b.pathHMACVerify(ctx, req, d, hmac)
+	case sigFound && hmacFound:
+		return logical.ErrorResponse("elements of batch_input must all provide 'signature' or all provide 'hmac'"), logical.ErrInvalidRequest
+
+	case missing && sigFound:
+		return logical.ErrorResponse("some elements of batch_input are missing 'signature'"), logical.ErrInvalidRequest
+
+	case missing && hmacFound:
+		return logical.ErrorResponse("some elements of batch_input are missing 'hmac'"), logical.ErrInvalidRequest
+
```

Comment:
Do you also need `case missing:`?  i.e. no elements provide either 'signature' or 'hmac'

Reply:
added case for missing on its own